src/utils/cosmos.ts

```diff
@@ -0,0 +1,151 @@
+        result.value.timeoutHeight.revisionNumber = Long.fromString(result.value.timeoutHeight.revisionNumber) as any
+      }
+      return result
+    }
+    return { ...m }
+  }
+
+  const msgsWithoutType = msgs.map(m => ({
+    ...manipulateMsg(m),
+    __type: undefined
+  }))
+
+  if (signType === 'AMINO') {
+    const signDoc = makeSignDoc(msgsWithoutType as any, fee as any, chainId, memo || undefined, account_number, sequence)
+    const signResponse = await provider.signAmino(chainId, cosmosTx.fromWalletAddress, signDoc)
+
+    const signedTx = cosmos.tx.v1beta1.TxRaw.encode({
+      bodyBytes: cosmos.tx.v1beta1.TxBody.encode({
+        messages: cosmosTx.data.protoMsgs.map(m => ({ type_url: m.type_url, value: new Uint8Array(m.value) })),
+        memo: signResponse.signed.memo
+      }).finish(),
+      authInfoBytes: cosmos.tx.v1beta1.AuthInfo.encode({
+        signerInfos: [
+          {
+            publicKey: {
+              type_url: '/cosmos.crypto.secp256k1.PubKey',
+              value: cosmos.crypto.secp256k1.PubKey.encode({
+                key: Buffer.from(
+                  signResponse.signature.pub_key.value,
+                  'base64'
+                )
+              }).finish()
+            },
+            modeInfo: {
+              single: {
+                mode: SignMode.SIGN_MODE_LEGACY_AMINO_JSON
+              }
+            },
+            sequence: Long.fromString(signResponse.signed.sequence)
+          }
+        ],
+        fee: {
+          amount: signResponse.signed.fee.amount as any[],
+          gasLimit: Long.fromString(signResponse.signed.fee.gas)
+        }
+      }).finish(),
+      signatures: [Buffer.from(signResponse.signature.signature, 'base64')]
+    }).finish()
+    const result = await provider.sendTx(chainId, signedTx, BroadcastMode.Async)
+    return uint8ArrayToHex(result)
+  } else if (signType === 'DIRECT') {
+    if (!rpcUrl) {
+      throw new Error('rpc url is undefined from server')
+    }
+
+    const sendingSigner = provider?.getOfflineSigner(chainId)
+    const sendingStargateClient = await SigningStargateClient?.connectWithSigner(
+      rpcUrl,
+      sendingSigner,
+      // @ts-expect-error
+      STARGATE_CLIENT_OPTIONS
+    )
+    const feeArray = fee?.amount[0] ? [{ denom: fee.amount[0].denom, amount: fee?.amount[0].amount }] : []
+
+    const isIbcTx = cosmosTx.data.msgs.filter(k => k.__type === 'DirectCosmosIBCTransferMessage').length > 0
+    const tmpGas = isIbcTx ? STARGATE_CLIENT_OPTIONS.gasLimits.ibcTransfer : STARGATE_CLIENT_OPTIONS.gasLimits.transfer
+
+    const broadcastTxRes = await sendingStargateClient.signAndBroadcast(
+      cosmosTx.fromWalletAddress,
+      msgs as any,
+      { gas: tmpGas.toString(), amount: feeArray }
+    )
+    return broadcastTxRes.transactionHash
+  } else {
+    throw new Error(`Sign type for cosmos not supported, type: ${signType}`)
+  }
+}
+
+export { getCosmosConnectedWallets, executeCosmosTransaction }
```

Comment:
Missed this file. Hopefully it was copied from a reliable source :D

Reply:
:D